// strong-agent/lib/profilers/memory.js
// Copyright IBM Corp. 2015. All Rights Reserved.
// Node module: strong-agent
// US Government Users Restricted Rights - Use, duplication or disclosure
// restricted by GSA ADP Schedule Contract with IBM Corp.

'use strict';

var debug = require('../debug')('profilers:memory');

function Instances() {
  this.addon = require('../addon');
  this.agent = null;
  this.enabled = false;
}
module.exports = new Instances;

Instances.prototype.init = function(agent_) {
  this.agent = agent_;
};

Instances.prototype.start = function() {
  if (!this.addon) {
    this.agent.info('could not start heap monitoring add-on');
    return false;
  }
  debug('instance monitoring started');
  this.addon.startHeapDiff();
  this.enabled = true;
  return true;
};

Instances.prototype.stop = function() {
  if (!this.enabled) return;
  debug('instance monitoring stopped');
  this.addon.stopHeapDiff(false);
  this.enabled = false;
};

Instances.prototype.poll = function() {
  if (!this.enabled) return null;
  var state = this.addon.stopHeapDiff(true);
  this.addon.startHeapDiff();
  return state;
};
